src/nnetbin/nnet-train-multistream.cc

```diff
@@ -0,0 +1,461 @@
+      Matrix<BaseFloat> feat_mat_host;
+      Posterior target_host;
+      Vector<BaseFloat> weight_host;
+      {
+        // Number of sequences (can have zero length),
+        int32 n_streams = num_streams;
+
+        // Create the final feature matrix with 'interleaved feature-lines',
+        feat_mat_host.Resize(n_streams * batch_size, nnet.InputDim(), kSetZero);
+        target_host.resize(n_streams * batch_size);
+        weight_host.Resize(n_streams * batch_size, kSetZero);
+        frame_num_utt.resize(n_streams, 0);
+
+        // we'll slice at most 'batch_size' frames,
+        for (int32 s = 0; s < n_streams; s++) {
+          int32 num_rows = feats_utt[s].NumRows();
+          frame_num_utt[s] = std::min(batch_size, num_rows);
+        }
+
+        // pack the data,
+        {
+          for (int32 s = 0; s < n_streams; s++) {
+            const Matrix<BaseFloat>& mat_tmp = feats_utt[s];
+            for (int32 r = 0; r < frame_num_utt[s]; r++) {
+              feat_mat_host.Row(r*n_streams + s).CopyFromVec(mat_tmp.Row(r));
+            }
+          }
+
+          for (int32 s = 0; s < n_streams; s++) {
+            const Posterior& target_tmp = labels_utt[s];
+            for (int32 r = 0; r < frame_num_utt[s]; r++) {
+              target_host[r*n_streams + s] = target_tmp[r];
+            }
+          }
+
+          // padded frames will keep initial zero-weight,
+          for (int32 s = 0; s < n_streams; s++) {
+            const Vector<BaseFloat>& weight_tmp = weights_utt[s];
+            for (int32 r = 0; r < frame_num_utt[s]; r++) {
+              weight_host(r*n_streams + s) = weight_tmp(r);
+            }
+          }
+        }
+
+        // remove the data we just packed,
+        {
+          for (int32 s = 0; s < n_streams; s++) {
+            // feats,
+            Matrix<BaseFloat>& m = feats_utt[s];
+            if (m.NumRows() == frame_num_utt[s]) {
+              feats_utt[s].Resize(0,0);  // we packed last chunk,
+            } else {
+              feats_utt[s] = Matrix<BaseFloat>(
+                m.RowRange(frame_num_utt[s], m.NumRows() - frame_num_utt[s])
+              );
+            }
+            // labels,
+            Posterior& post = labels_utt[s];
+            post.erase(post.begin(), post.begin() + frame_num_utt[s]);
+            // weights,
+            Vector<BaseFloat>& w = weights_utt[s];
+            if (w.Dim() == frame_num_utt[s]) {
+              weights_utt[s].Resize(0);  // we packed last chunk,
+            } else {
+              weights_utt[s] = Vector<BaseFloat>(
+                w.Range(frame_num_utt[s], w.Dim() - frame_num_utt[s])
+              );
+            }
+          }
+        }
+      }
+
+      // pass the info about padding,
+      nnet.SetSeqLengths(frame_num_utt);
+      // Show the 'utt' lengths in the VLOG[2],
+      if (kaldi::g_kaldi_verbose_level >= 2) {
+        WriteIntegerVector(
+          KALDI_LOG << "frame_num_utt[" << frame_num_utt.size() << "]",
```

Comment:
Karel, I don't know what's supposed to be going on here.. I'm astonished that this would compile, but even if it does, it's not OK.  The macros like KALDI_LOG are supposed to just be invoked, there should be no assumption about what they return.  You may have to use std::ostringstream.